File: server/services/tournamentService.js
const Tournament = require("../models/Tournament");
const Team = require("../models/Team");
const MatchResult = require("../models/MatchResult");
const PrizeDistribution = require("../models/PrizeDistribution");
const Notification = require("../models/Notification");

const ApiError = require("../utils/ApiError");
const { ROLES } = require("../constants/roles");

const isPlatformAdmin = (user) => {
  return user.role === ROLES.ADMIN || user.role === ROLES.SUPER_ADMIN;
};

const isTournamentOwner = (tournament, user) => {
  if (!tournament.createdBy) return false;

  return tournament.createdBy.toString() === user._id.toString();
};

const ensureCanManageTournament = (tournament, user) => {
  if (isPlatformAdmin(user)) return;

  if (user.role === ROLES.ORGANIZER && isTournamentOwner(tournament, user)) {
    return;
  }

  throw new ApiError(403, "You can manage only your own tournaments");
};

const notifyTournamentPlayers = async (tournament, title, message) => {
  const teams = await Team.find({
    _id: { $in: tournament.registeredTeams },
  });

  const userIds = new Set();

  teams.forEach((team) => {
    team.players.forEach((playerId) => {
      userIds.add(playerId.toString());
    });

    userIds.add(team.igl.toString());
  });

  const notifications = Array.from(userIds).map((userId) => ({
    user: userId,
    title,
    message,
    type: "tournament",
  }));

  if (notifications.length > 0) {
    await Notification.insertMany(notifications);
  }
};

const createTournament = async (tournamentData, currentUser) => {
  const { title, game, mode, entryFee, prizePool, maxTeams, startDate } =
    tournamentData;

  if (!title || !game || !mode || !startDate) {
    throw new ApiError(400, "Please fill all required fields");
  }

  if (entryFee !== undefined && Number.isNaN(Number(entryFee))) {
    throw new ApiError(400, "Entry fee must be a number");
  }

  if (prizePool !== undefined && Number.isNaN(Number(prizePool))) {
    throw new ApiError(400, "Prize pool must be a number");
  }

  if (maxTeams !== undefined && Number.isNaN(Number(maxTeams))) {
    throw new ApiError(400, "Max teams must be a number");
  }

  if (Number(entryFee) < 0 || Number(prizePool) < 0) {
    throw new ApiError(400, "Entry fee and prize pool cannot be negative");
  }

  if (maxTeams !== undefined && Number(maxTeams) <= 0) {
    throw new ApiError(400, "Max teams must be greater than 0");
  }

  const existingTournament = await Tournament.findOne({
    title: title.trim(),
  });

  if (existingTournament) {
    throw new ApiError(400, "Tournament already exists");
  }

  const tournament = await Tournament.create({
    title: title.trim(),
    game,
    mode,
    entryFee: Number(entryFee) || 0,
    prizePool: Number(prizePool) || 0,
    maxTeams: Number(maxTeams) || 25,
    startDate,
    createdBy: currentUser._id,
  });

  return tournament;
};

const registerTeam = async (tournamentId, currentUser) => {
  const tournament = await Tournament.findById(tournamentId);

  if (!tournament) {
    throw new ApiError(404, "Tournament not found");
  }

  if (tournament.status === "completed") {
    throw new ApiError(400, "Cannot register in completed tournament");
  }

  const team = await Team.findOne({
    players: currentUser._id,
  });

  if (!team) {
    throw new ApiError(400, "You are not in any team");
  }

  const alreadyRegistered = tournament.registeredTeams.some(
    (teamId) => teamId.toString() === team._id.toString(),
  );

  if (alreadyRegistered) {
    throw new ApiError(400, "Team already registered");
  }

  if (tournament.registeredTeams.length >= tournament.maxTeams) {
    throw new ApiError(400, "Tournament is full");
  }

  tournament.registeredTeams.push(team._id);
  await tournament.save();

  return tournament;
};

const getAllTournaments = async () => {
  const tournaments = await Tournament.find()
    .populate("createdBy", "name email role")
    .sort({ createdAt: -1 });

  return tournaments;
};

const getMyCreatedTournaments = async (currentUser) => {
  const query = isPlatformAdmin(currentUser)
    ? {}
    : {
        createdBy: currentUser._id,
      };

  const tournaments = await Tournament.find(query)
    .populate("createdBy", "name email role")
    .sort({ createdAt: -1 });

  return tournaments;
};

const getTournamentById = async (tournamentId) => {
  const tournament = await Tournament.findById(tournamentId)
    .populate("registeredTeams")
    .populate("createdBy", "name email role");

  if (!tournament) {
    throw new ApiError(404, "Tournament not found");
  }

  return tournament;
};

const leaveTournament = async (tournamentId, currentUser) => {
  const tournament = await Tournament.findById(tournamentId);

  if (!tournament) {
    throw new ApiError(404, "Tournament not found");
  }

  if (tournament.status === "live") {
    throw new ApiError(400, "Cannot leave tournament after it has started");
  }

  if (tournament.status === "completed") {
    throw new ApiError(400, "Cannot leave completed tournament");
  }

  const team = await Team.findOne({
    players: currentUser._id,
  });

  if (!team) {
    throw new ApiError(400, "You are not in any team");
  }

  const isRegistered = tournament.registeredTeams.some(
    (teamId) => teamId.toString() === team._id.toString(),
  );

  if (!isRegistered) {
    throw new ApiError(400, "Team is not registered in this tournament");
  }

  tournament.registeredTeams.pull(team._id);
  await tournament.save();

  return tournament;
};

const startTournament = async (tournamentId, currentUser) => {
  const tournament = await Tournament.findById(tournamentId);

  if (!tournament) {
    throw new ApiError(404, "Tournament not found");
  }

  ensureCanManageTournament(tournament, currentUser);

  if (tournament.status === "live") {
    throw new ApiError(400, "Tournament is already live");
  }

  if (tournament.status === "completed") {
    throw new ApiError(400, "Tournament is already completed");
  }

  if (tournament.registeredTeams.length === 0) {
    throw new ApiError(400, "No teams registered");
  }

  tournament.status = "live";
  await tournament.save();

  await notifyTournamentPlayers(
    tournament,
    "Tournament Started",
    `${tournament.title} has started. Join your match room on time.`,
  );

  return tournament;
};

const completeTournament = async (tournamentId, currentUser) => {
  const tournament = await Tournament.findById(tournamentId);

  if (!tournament) {
    throw new ApiError(404, "Tournament not found");
  }

  ensureCanManageTournament(tournament, currentUser);

  if (tournament.status === "completed") {
    throw new ApiError(400, "Tournament already completed");
  }

  const results = await MatchResult.find({
    tournament: tournamentId,
  }).populate("team", "teamName");

  if (results.length === 0) {
    throw new ApiError(400, "No match results found");
  }

  const leaderboard = {};

  results.forEach((result) => {
    const teamId = result.team._id.toString();

    if (!leaderboard[teamId]) {
      leaderboard[teamId] = {
        team: result.team,
        points: 0,
      };
    }

    leaderboard[teamId].points += result.totalPoints;
  });

  const sortedTeams = Object.values(leaderboard).sort(
    (a, b) => b.points - a.points,
  );

  const prizePool = tournament.prizePool;

  const firstPrize = Math.floor(prizePool * 0.5);
  const secondPrize = Math.floor(prizePool * 0.3);
  const thirdPrize = Math.floor(prizePool * 0.2);

  await PrizeDistribution.create({
    tournament: tournamentId,

    firstPlace: {
      team: sortedTeams[0]?.team?._id,
      amount: firstPrize,
    },

    secondPlace: {
      team: sortedTeams[1]?.team?._id,
      amount: secondPrize,
    },

    thirdPlace: {
      team: sortedTeams[2]?.team?._id,
      amount: thirdPrize,
    },
  });

  tournament.status = "completed";
  await tournament.save();

  await notifyTournamentPlayers(
    tournament,
    "Tournament Completed",
    `${tournament.title} has been completed. Winner: ${sortedTeams[0]?.team?.teamName}`,
  );

  return {
    tournament,
    winner: sortedTeams[0]?.team?.teamName,
    prizePool,
  };
};

const getTournamentHistory = async () => {
  const tournaments = await Tournament.find({
    status: "completed",
  }).sort({ updatedAt: -1 });

  const history = [];

  for (const tournament of tournaments) {
    const prize = await PrizeDistribution.findOne({
      tournament: tournament._id,
    })
      .populate("firstPlace.team", "teamName")
      .populate("secondPlace.team", "teamName")
      .populate("thirdPlace.team", "teamName");

    history.push({
      tournamentId: tournament._id,
      tournamentName: tournament.title,
      game: tournament.game,
      mode: tournament.mode,
      prizePool: tournament.prizePool,
      entryFee: tournament.entryFee,
      totalRegisteredTeams: tournament.registeredTeams.length,

      winner: prize?.firstPlace?.team?.teamName || "No winner found",

      firstPlace: {
        team: prize?.firstPlace?.team?.teamName || null,
        amount: prize?.firstPlace?.amount || 0,
      },

      secondPlace: {
        team: prize?.secondPlace?.team?.teamName || null,
        amount: prize?.secondPlace?.amount || 0,
      },

      thirdPlace: {
        team: prize?.thirdPlace?.team?.teamName || null,
        amount: prize?.thirdPlace?.amount || 0,
      },

      completedAt: tournament.updatedAt,
    });
  }

  return history;
};

module.exports = {
  createTournament,
  registerTeam,
  getAllTournaments,
  getMyCreatedTournaments,
  getTournamentById,
  leaveTournament,
  startTournament,
  completeTournament,
  getTournamentHistory,
};

File: backend/app/services/tournament_service.py
from decimal import Decimal

from fastapi import HTTPException
from sqlalchemy import select, func
from sqlalchemy.orm import Session

from app.models.domain import (
    Tournament,
    TournamentRegistration,
    Team,
    TeamMember,
    MatchRoom,
    MatchResult,
    PrizeDistribution,
    Notification,
)
from app.constants.roles import Role
from app.constants.status import TournamentStatus, NotificationType
from app.services.notification_service import create as notify
from app.utils.serializers import tournament_dict


def can_manage(t, u):
    if u.role in {Role.ADMIN, Role.SUPER_ADMIN}:
        return True

    return u.role == Role.ORGANIZER and t.created_by == u.id


def create(db, u, d):
    if u.role not in {Role.ORGANIZER, Role.ADMIN, Role.SUPER_ADMIN}:
        raise HTTPException(403, "Access Denied")

    if db.scalar(
        select(Tournament).where(
            Tournament.title == d.title.strip()
        )
    ):
        raise HTTPException(400, "Tournament already exists")

    if d.startDate is None:
        raise HTTPException(400, "Start date is required")

    t = Tournament(
        title=d.title.strip(),
        game=d.game,
        mode=d.mode,
        entry_fee=d.entryFee,
        prize_pool=d.prizePool,
        max_teams=d.maxTeams,
        start_date=d.startDate,
        created_by=u.id,
    )

    db.add(t)
    db.commit()
    db.refresh(t)

    return t


def list_all(db):
    return db.scalars(
        select(Tournament).order_by(Tournament.created_at.desc())
    ).unique().all()


def my_created(db, u):
    q = select(Tournament).order_by(Tournament.created_at.desc())

    if u.role not in {Role.ADMIN, Role.SUPER_ADMIN}:
        q = q.where(Tournament.created_by == u.id)

    return db.scalars(q).unique().all()


def get(db, tid):
    t = db.get(Tournament, tid)

    if not t:
        raise HTTPException(404, "Tournament not found")

    return t


def register(db, u, tid):
    t = get(db, tid)

    if t.status == TournamentStatus.COMPLETED:
        raise HTTPException(
            400,
            "Cannot register in completed tournament",
        )

    team = db.scalar(
        select(Team)
        .join(TeamMember)
        .where(TeamMember.user_id == u.id)
    )

    if not team:
        raise HTTPException(
            400,
            "You are not in any team",
        )

    if db.scalar(
        select(TournamentRegistration).where(
            TournamentRegistration.tournament_id == t.id,
            TournamentRegistration.team_id == team.id,
        )
    ):
        raise HTTPException(
            400,
            "Team already registered",
        )

    count = db.scalar(
        select(func.count())
        .select_from(TournamentRegistration)
        .where(
            TournamentRegistration.tournament_id == t.id
        )
    )

    if count >= t.max_teams:
        raise HTTPException(
            400,
            "Tournament is full",
        )

    db.add(
        TournamentRegistration(
            tournament_id=t.id,
            team_id=team.id,
        )
    )

    db.commit()
    db.refresh(t)

    return t


def leave(db, u, tid):
    t = get(db, tid)

    if t.status == TournamentStatus.LIVE:
        raise HTTPException(
            400,
            "Cannot leave tournament after it has started",
        )

    if t.status == TournamentStatus.COMPLETED:
        raise HTTPException(
            400,
            "Cannot leave completed tournament",
        )

    team = db.scalar(
        select(Team)
        .join(TeamMember)
        .where(TeamMember.user_id == u.id)
    )

    if not team:
        raise HTTPException(
            400,
            "You are not in any team",
        )

    r = db.scalar(
        select(TournamentRegistration).where(
            TournamentRegistration.tournament_id == t.id,
            TournamentRegistration.team_id == team.id,
        )
    )

    if not r:
        raise HTTPException(
            400,
            "Team is not registered in this tournament",
        )

    db.delete(r)
    db.commit()
    db.refresh(t)

    return t


def start(db, u, tid):
    t = get(db, tid)

    if not can_manage(t, u):
        raise HTTPException(
            403,
            "You can manage only your own tournaments",
        )

    if t.status == TournamentStatus.LIVE:
        raise HTTPException(
            400,
            "Tournament is already live",
        )

    if t.status == TournamentStatus.COMPLETED:
        raise HTTPException(
            400,
            "Tournament is already completed",
        )

    if not t.registrations:
        raise HTTPException(
            400,
            "No teams registered",
        )

    t.status = TournamentStatus.LIVE

    for r in t.registrations:
        for m in r.team.members:
            notify(
                db,
                m.user_id,
                "Tournament Started",
                f"{t.title} has started. Join your match room on time.",
                NotificationType.TOURNAMENT,
            )

    db.commit()
    db.refresh(t)

    return t


def complete(db, u, tid):
    t = get(db, tid)

    if not can_manage(t, u):
        raise HTTPException(
            403,
            "You can manage only your own tournaments",
        )

    if t.status == TournamentStatus.COMPLETED:
        raise HTTPException(
            400,
            "Tournament already completed",
        )

    results = db.scalars(
        select(MatchResult).where(
            MatchResult.tournament_id == t.id
        )
    ).all()

    if not results:
        raise HTTPException(
            400,
            "No match results found",
        )

    totals = {}

    for r in results:
        totals[r.team_id] = (
            totals.get(r.team_id, 0) + r.total_points
        )

    sorted_teams = sorted(
        totals.items(),
        key=lambda x: x[1],
        reverse=True,
    )

    # Prize distribution
    # PostgreSQL NUMERIC values are returned by SQLAlchemy
    # as Decimal objects. Therefore, use Decimal values
    # instead of Python floats for money calculations.
    pool = t.prize_pool

    first = pool * Decimal("0.50")
    second = pool * Decimal("0.30")
    third = pool * Decimal("0.20")

    old = db.scalar(
        select(PrizeDistribution).where(
            PrizeDistribution.tournament_id == t.id
        )
    )

    if old:
        db.delete(old)
        db.flush()

    prize = PrizeDistribution(
        tournament_id=t.id,
        first_team_id=(
            sorted_teams[0][0]
            if len(sorted_teams) > 0
            else None
        ),
        first_amount=first,
        second_team_id=(
            sorted_teams[1][0]
            if len(sorted_teams) > 1
            else None
        ),
        second_amount=second,
        third_team_id=(
            sorted_teams[2][0]
            if len(sorted_teams) > 2
            else None
        ),
        third_amount=third,
    )

    db.add(prize)

    t.status = TournamentStatus.COMPLETED

    winner_team = (
        db.get(Team, sorted_teams[0][0])
        if sorted_teams
        else None
    )

    for r in t.registrations:
        for m in r.team.members:
            notify(
                db,
                m.user_id,
                "Tournament Completed",
                (
                    f"{t.title} has been completed. "
                    f"Winner: "
                    f"{winner_team.team_name if winner_team else 'N/A'}"
                ),
                NotificationType.TOURNAMENT,
            )

    db.commit()
    db.refresh(t)

    return {
        "tournament": t,
        "winner": (
            winner_team.team_name
            if winner_team
            else None
        ),
        "prizePool": float(pool),
    }


def update(db, u, tid, d):
    t = get(db, tid)

    if not can_manage(t, u):
        raise HTTPException(
            403,
            "You can manage only your own tournaments",
        )

    if t.status == TournamentStatus.COMPLETED:
        raise HTTPException(
            400,
            "Completed tournament cannot be updated",
        )

    if (
        d.title is not None
        and d.title.strip() != t.title
        and db.scalar(
            select(Tournament).where(
                Tournament.title == d.title.strip()
            )
        )
    ):
        raise HTTPException(
            400,
            "Tournament title already exists",
        )

    if d.title is not None:
        t.title = d.title.strip()

    if d.game is not None:
        t.game = d.game

    if d.mode is not None:
        t.mode = d.mode

    if d.entryFee is not None:
        t.entry_fee = d.entryFee

    if d.prizePool is not None:
        t.prize_pool = d.prizePool

    if d.maxTeams is not None:
        if d.maxTeams < len(t.registrations):
            raise HTTPException(
                400,
                "Max teams cannot be less than already registered teams",
            )

        t.max_teams = d.maxTeams

    if d.startDate is not None:
        t.start_date = d.startDate

    db.commit()
    db.refresh(t)

    return t


def delete_tournament(db, u, tid):
    t = get(db, tid)

    if not can_manage(t, u):
        raise HTTPException(
            403,
            "You can manage only your own tournaments",
        )

    db.delete(t)
    db.commit()


def history(db):
    ts = db.scalars(
        select(Tournament)
        .where(
            Tournament.status == TournamentStatus.COMPLETED
        )
        .order_by(Tournament.updated_at.desc())
    ).all()

    out = []

    for t in ts:
        p = db.scalar(
            select(PrizeDistribution).where(
                PrizeDistribution.tournament_id == t.id
            )
        )

        out.append(
            {
                "tournamentId": str(t.id),
                "tournamentName": t.title,
                "game": t.game,
                "mode": t.mode,
                "prizePool": float(t.prize_pool),
                "entryFee": float(t.entry_fee),
                "totalRegisteredTeams": len(t.registrations),
                "winner": (
                    str(p.first_team.team_name)
                    if p and p.first_team
                    else "No winner found"
                ),
                "firstPlace": {
                    "team": (
                        p.first_team.team_name
                        if p and p.first_team
                        else None
                    ),
                    "amount": (
                        float(p.first_amount)
                        if p
                        else 0
                    ),
                },
                "secondPlace": {
                    "team": (
                        p.second_team.team_name
                        if p and p.second_team
                        else None
                    ),
                    "amount": (
                        float(p.second_amount)
                        if p
                        else 0
                    ),
                },
                "thirdPlace": {
                    "team": (
                        p.third_team.team_name
                        if p and p.third_team
                        else None
                    ),
                    "amount": (
                        float(p.third_amount)
                        if p
                        else 0
                    ),
                },
                "completedAt": t.updated_at,
            }
        )

    return out